cvpaperchallenge/Ascender
# 68/feat/use cache in CI job

This is still work in progress PR.

## Issue URL

https://github.com/cvpaperchallenge/Ascender/issues/68

## Change overview

### Run `poetry install` inside of Docker image build step at CI job
`poetry install` is time consuming step. So if we cache this step, we can make reduce CI job much faster. To achieve this following points are changed:

-  A new ARG `RUN_POETRY_INSTALL_AT_BUILD_TIME` is added to `Dockerfile`, and if this is true, run `poetry install` at Docker image built step. A `.venv` directory created by above `poetry install` is moved from `APPLICATION_DIRECTORY` to prevent overwrite by volume operation of Docker (This moved `.venv` directory is moved back to `APPLICATION_DIRECTORY` inside of CI job.).

- A new step `Move .venv directory` is added to CI job `lint-and-test` to move back `.venv` directory. And `Install dependencies` step is removed from CI job because it is already done in Docker image built step.


### Introduce cache to CI job to cache Docker image
As suggested in #68, currently our CI job doesn't have any cache system. So CI job is taking long time. To mitigate this, followings are introduced:

- To accelerate Docker image build process, layer level cache is introduced. We used [`buildx bake`](https://docs.docker.com/engine/reference/commandline/buildx_bake/) to build with cache. Cache directory is created by official action [`actions/cache@v3`](https://github.com/actions/cache). Note that, an action [`build-push-action
`](https://github.com/docker/build-push-action) supports build with cache. However, this action doesn't support input from `docker-compose.yaml`. So I decided to use `buildx bake` instead.

- To cache built image, local registry is introduced. Before running `docker compose up`, built image is pulled from local registry. To push local registry, we need to appropriately specify image name and tag. To achieve this, `environments/ci/docker-compose.yaml` is added. This feature is not well tested yet, so I added `USE_CACHE` environmental variable to `lint-and-test.yaml` and if this is not `true`, CI job will be run without cache.

## How to test

Please check CI result.

First time CI job run (no cache): 2min 54s
https://github.com/cvpaperchallenge/Ascender/actions/runs/3257278867/jobs/5348444758
![Screenshot from 2022-10-16 07-00-46](https://user-images.githubusercontent.com/21344007/196011114-1c5fe0c4-6fd0-4866-acf1-1eefdf3a5181.png)


Second time CI job run (first time cache is used): 1min 17s
https://github.com/cvpaperchallenge/Ascender/actions/runs/3257278867/jobs/5348469025
![Screenshot from 2022-10-16 07-03-56](https://user-images.githubusercontent.com/21344007/196011110-9ed36cd8-83bc-496b-8d30-f9891ab3c661.png)

## Note for reviewers

There is no way to test CI job at local PC. So I did many try and error by actually pushing the code. As a result, many commit logs of these process are left and I'm sorry about this.
